impl/maven-core/src/main/java/org/apache/maven/lifecycle/internal/builder/multithreaded/SmartProjectComparator.java

```diff
@@ -80,7 +80,18 @@ public Comparator<MavenProject> getComparator() {
      * @return the project's weight (higher means longer dependency chain)
      */
     public long getProjectWeight(MavenProject project) {
-        return projectWeights.computeIfAbsent(project, this::calculateWeight);
+        // First check if weight is already calculated
+        Long existingWeight = projectWeights.get(project);
+        if (existingWeight != null) {
+            return existingWeight;
+        }
+
+        // Calculate weight without using computeIfAbsent to avoid recursive update issues
+        long weight = calculateWeight(project);
```

Comment:
should we synchronize this call or method to avoid concurrent computation, or it is cheap computation and not need synchronize?

Reply:
I think it's lightweight enough and I'd rather not synchronize the whole SmartProjectComparator.